WIP: Fixes #589. Allows 2+ roles to delegate to same role

awwad opened the pull request:
Removes an incorrect check that prevents delegating to a role if any role has previously delegated to that role. (Previously, if X delegated to A, then Y was prevented from delegating to A.)

Adds a few tests.

See #589 for more details. (Note that the central fix, a removal of a few lines, has been merged already in #638, and so has been removed from this commit history.)

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

vladimir-v-diaz commented:
Can you please edit the corresponding test case?https://github.com/theupdateframework/tuf/blob/aa2ab218f22d8682e03c992ea98f88efd155cffd/tests/test_repository_tool.py#L1192-L1196

vladimir-v-diaz commented on tuf/repository_tool.py:
```diff
@@ -2222,11 +2222,6 @@ def delegate(self, rolename, public_keys, paths, threshold=1,
     if path_hash_prefixes is not None:
       securesystemslib.formats.PATH_HASH_PREFIXES_SCHEMA.check_match(path_hash_prefixes)
 
-    # Check if 'rolename' is not already a delegation.
```
It might be good to mention in the docstring, or in a comment, that 2+ roles can delegate to a single role.
https://github.com/awwad/tuf/blob/2e590485e540b822117de76d1ec949731720ba6f/tuf/repository_tool.py#L2134-L2144

awwad replied:
Actually, this doesn't feel natural. There's no reason to suspect that 2+ roles couldn't delegate to a single role in the code, docstring, or comments. I'll take a look at the documentation - maybe there's somewhere to note that there.

awwad commented:
Will do

awwad commented:
Rebased to remove conflicts tidily.

awwad commented:
Fixed the test case to instead expect success, but we have to add some more detailed testing that covers these delegation scenarios, and until we do, this is still WIP.

awwad commented:
Some tests to add before this PR is OK:

### Scenario 1:
Targets delegates `"*"` to A, which delegates `"a*"` to C
Targets delegates `"*"` to B, which delegates `"b*"` to C
All roles are signed by all keys expected of them. (e.g. C is signed by a threshold of all keys that A and B expect of it for the delegations listed)
C lists targets `a1.zip` and `b1.zip`.

Tests needed:
- [x] all delegations above must succeed
- [ ] `get_one_valid_targetinfo(b1.zip)` should successfully fetch the fileinfo C listed for `b1.zip`
- [ ] `get_one_valid_targetinfo(a1.zip)` should successfully fetch the fileinfo C listed for `a1.zip`

### Scenario 2:
Same as Scenario 1, except that role C is NOT signed by the keys expected of it by role A (but still is signed by the keys expected of it by role B). (This is important to avoid a particular "attack" whose explanation I'm sitting on and will get to in a fresh issue later.)

Tests needed:
- [ ] `get_one_valid_targetinfo(b1.zip)` should successfully fetch the fileinfo C listed for `b1.zip`
- [ ] `get_one_valid_targetinfo(a1.zip)` should of course fail to fetch the fileinfo C listed for `a1.zip`

I may add some tests to this when I go back over my notes later.

awwad commented:
Will rebase and conflict resolve this due to merging of #638, which performs part of what this pull request was intended to do. This PR will then add the tests I recommended.

awwad commented:
Rebased. The check I removed here, which you also removed in the later-merged #638, is no longer in the commit history here. I've changed the docs changes as well: while the changes were previously in tuf/README.md, that has been split into docs/CLI.md and docs/TUTORIAL.md, and so the same notes as were added before to tuf/README.md have now been added to those two documents. You may want to take a peek at the changed docs/CLI.md and make sure the extra note isn't too disruptive.

vladimir-v-diaz commented:
I don't find the changes to `CLI.md` too disruptive.

vladimir-v-diaz commented:
@awwad 
I want to validate the correctness of the API redesign against this pull request, if it makes sense.

Has the task/test in scenario 2 (https://github.com/theupdateframework/tuf/pull/590#issuecomment-359841774) been implemented in this pull request?  I see that Travis is failing, but that is expected, correct?

awwad commented:
The checks in the comment capture the current state -- the last two tests in the list still need to be added. There are more tests I'd want for the API redesign, I think, though I'd have to make sure I understand what you mean by that: is the redesign the fix for #660? (For example, is the expectation that the redesign will result in validation of a delegated role to always consult the appropriate delegating role to determine the keys to expect?)

vladimir-v-diaz commented:
> is the redesign the fix for #660? (For example, is the expectation that the redesign will result in validation of a delegated role to always consult the appropriate delegating role to determine the keys to expect?)

That's correct!  The redesign of the API is outlined in https://github.com/theupdateframework/tuf/issues/574#issuecomment-358115382 and being worked on in https://github.com/vladimir-v-diaz/tuf/tree/refactor_low_level_api.

vladimir-v-diaz commented:
@awwad Is it okay if I finish the last two tests in https://github.com/theupdateframework/tuf/pull/590#issuecomment-359841774?  I'd like to use this pull request to test https://github.com/theupdateframework/tuf/pull/757.

I can let you know when pull request #757 is ready for a review.

awwad commented:
Certainly

joshuagl commented:
Due to imminent refactor efforts this code is unlikely to merge.
